File: patterns/Creational/AbstractFactory/Go/RealWorldExample/DataBaseConnection.go
package main

import "fmt"

// Abstract Factory interface for creating database connections and queries
type DatabaseFactory interface {
	CreateConnection() Connection
	CreateQuery() Query
}

// Concrete Factory for MySQL database
type MySQLFactory struct{}

func (f *MySQLFactory) CreateConnection() Connection {
	return &MySQLConnection{}
}

func (f *MySQLFactory) CreateQuery() Query {
	return &MySQLQuery{}
}

// Concrete Factory for PostgreSQL database
type PostgreSQLFactory struct{}

func (f *PostgreSQLFactory) CreateConnection() Connection {
	return &PostgreSQLConnection{}
}

func (f *PostgreSQLFactory) CreateQuery() Query {
	return &PostgreSQLQuery{}
}

// Abstract product interface for database connections
type Connection interface {
	Connect() string
}

// Concrete product for MySQL connection
type MySQLConnection struct{}

func (c *MySQLConnection) Connect() string {
	return "Connected to MySQL database."
}

// Concrete product for PostgreSQL connection
type PostgreSQLConnection struct{}

func (c *PostgreSQLConnection) Connect() string {
	return "Connected to PostgreSQL database."
}

// Abstract product interface for database queries
type Query interface {
	Execute() string
}

// Concrete product for MySQL query
type MySQLQuery struct{}

func (q *MySQLQuery) Execute() string {
	return "Executing MySQL query."
}

// Concrete product for PostgreSQL query
type PostgreSQLQuery struct{}

func (q *PostgreSQLQuery) Execute() string {
	return "Executing PostgreSQL query."
}

// Client code that interacts with the abstract factory and products
func clientCode(factory DatabaseFactory) {
	connection := factory.CreateConnection()
	fmt.Println(connection.Connect())

	query := factory.CreateQuery()
	fmt.Println(query.Execute())
}

func main() {
	// Testing MySQL factory
	fmt.Println("Testing MySQL factory:")
	clientCode(&MySQLFactory{})
	fmt.Println()

	// Testing PostgreSQL factory
	fmt.Println("Testing PostgreSQL factory:")
	clientCode(&PostgreSQLFactory{})
}

File: patterns/Creational/AbstractFactory/JavaScript/RealWorldExample/DataBaseConnection.js
// AbstractFactory.js - DatabaseFactory Interface
class DatabaseFactory {
    createConnection() {
      throw new Error("Method 'createConnection()' must be implemented.");
    }
  
    createQuery() {
      throw new Error("Method 'createQuery()' must be implemented.");
    }
  }
  
  // MySQLFactory.js - Concrete Factory for MySQL
  class MySQLFactory extends DatabaseFactory {
    createConnection() {
      return new MySQLConnection();
    }
  
    createQuery() {
      return new MySQLQuery();
    }
  }
  
  // PostgreSQLFactory.js - Concrete Factory for PostgreSQL
  class PostgreSQLFactory extends DatabaseFactory {
    createConnection() {
      return new PostgreSQLConnection();
    }
  
    createQuery() {
      return new PostgreSQLQuery();
    }
  }
  
  // Connection.js - Connection Interface
  class Connection {
    connect() {
      throw new Error("Method 'connect()' must be implemented.");
    }
  }
  
  // MySQLConnection.js - Concrete MySQL Connection
  class MySQLConnection extends Connection {
    connect() {
      return "Connected to MySQL database.";
    }
  }
  
  // PostgreSQLConnection.js - Concrete PostgreSQL Connection
  class PostgreSQLConnection extends Connection {
    connect() {
      return "Connected to PostgreSQL database.";
    }
  }
  
  // Query.js - Query Interface
  class Query {
    execute() {
      throw new Error("Method 'execute()' must be implemented.");
    }
  }
  
  // MySQLQuery.js - Concrete MySQL Query
  class MySQLQuery extends Query {
    execute() {
      return "Executing MySQL query.";
    }
  }
  
  // PostgreSQLQuery.js - Concrete PostgreSQL Query
  class PostgreSQLQuery extends Query {
    execute() {
      return "Executing PostgreSQL query.";
    }
  }
  
  // Client.js - Client Code
  function clientCode(factory) {
    const connection = factory.createConnection();
    console.log(connection.connect());
  
    const query = factory.createQuery();
    console.log(query.execute());
  }
  
  // Usage
  console.log("Testing MySQL factory:");
  clientCode(new MySQLFactory());
  console.log("\n");
  
  console.log("Testing PostgreSQL factory:");
  clientCode(new PostgreSQLFactory());
  